### PR TITLE
Auto: implement issue #503

### LINKED ISSUE #503: Add WebSocket auth handshake router in src/api/routers/websocket.py
Subtask of #489
Depends on #502

Create src/api/routers/websocket.py with a FastAPI WebSocket router. On handshake, validate the JWT token from the query param or header using the same auth utilities as other routers. Reject with 401 on invalid/missing token. After auth, accept the connection and route to a channel based on path params resource_type and resource_id (e.g. /ws/{resource_type}/{resource_id}). ConnectionManager.join is called on accept, ConnectionManager.leave on disconnect. Wire the router into main.py under the /ws prefix. No broadcast logic here — just auth + join/leave. Write unit tests for the handshake: valid token accepts, missing token rejects, invalid token rejects with 401.

Dev-Plan: docs/dev-plan/00-foundations/0503-add-websocket-auth-handshake-router-in-src-api-routers-webso.md

### PR DESCRIPTION
**Pull request by yihua-zhuo**
Closes #503

Plan: `.plans/issue-503.md`

This PR was auto-implemented by Claude Code in three stages:

1. **Plan** — Claude wrote a detailed implementation plan to `.plans/issue-503.md`.
2. **Implement** — Claude executed the plan, ran `ruff check src/` + the unit and integration test suites, and iterated until all three were green.
3. **Self-review** — Claude reviewed its own diff against `codereview.md` (the project review checklist) and fixed any rule violations found, re-running the test suite if any changes were made.

PR-time CI (`Dev Agent CI`) re-verifies independently before this PR is mergeable.

Please review before merging.

### CODE REVIEW

File: src/api/routers/websocket.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    removed from the channel.
+    """
+    ctx = verify_ws_token(websocket)
+    if ctx is None:
+        await websocket.close(code=WS_1008_POLICY_VIOLATION, reason="Unauthorized")
+        return
+
+    await websocket.accept()
+
+    channel = f"{ctx.tenant_id}:{resource_type}:{resource_id}"
```
Comment (yihua-zhuo) [inline src/api/routers/websocket.py:61]: **CRITICAL** (critical)

**Issue:** Rule 126: channel key is built with no guard for `tenant_id=None`; an authenticated connection with a null tenant produces the key `None:resource_type:resource_id`, breaking isolation

**Suggestion:** Guard against None tenant_id: if ctx.tenant_id is None, close with code 1008 and reason 'Invalid tenant context' before building the channel key

File: .plans/issue-503.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+from .websocket import router
+__all__ = ["router"]
+```
+This makes the router discoverable by `iter_routers()` in `src/api/__init__.py` without any change to `main.py` — the existing `for router in iter_routers(): app.include_router(router)` loop already picks up all `APIRouter` exports in the package.
+
+**Complete判定**: `PYTHONPATH=src python -c "from api.routers.websocket import router; print('import ok')"` → no `ImportError`.
+
+**Step 4: Write `tests/unit/test_websocket.py`**
+
+Create the test file with `verify_ws_token` tests and integration-style WebSocket connection tests using `fastapi.testclient.TestClient`:
```
Comment (yihua-zhuo) [inline .plans/issue-503.md:85]: **INFO** (suggestion)

**Issue:** Plan lists tests as 'unit tests' but Step 4 specifies TestClient.websocket_connect (integration-style) — test category label is ambiguous though case count (6) is correct

**Suggestion:** Clarify whether tests/unit/test_websocket.py uses TestClient (integration style) or raw AsyncMock; label should match the chosen approach

File: src/api/routers/websocket.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        auth_header = websocket.headers.get("authorization", "")
+        if auth_header.startswith("Bearer "):
+            raw_token = auth_header[7:]
+
+    if not raw_token:
+        return None
+
+    try:
+        return _decode_jwt(raw_token)
+    except Exception:
```
Comment (yihua-zhuo) [inline src/api/routers/websocket.py]: **WARNING** (suggestion)

**Issue:** Rule 25: bare `except Exception` swallows all errors, not just JWT decode failures — unexpected attribute errors on `websocket` will be silently absorbed

**Suggestion:** Catch only `HTTPException` since that is what _decode_jwt raises on token errors; let other exceptions propagate: except HTTPException: return None

File: src/api/routers/websocket.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        return None
+
+
+@router.websocket("/{resource_type}/{resource_id}")
```
Comment (yihua-zhuo) [inline src/api/routers/websocket.py:38]: **INFO** (suggestion)

**Issue:** Rule 15: no validation that resource_type/resource_id are non-empty or safe for channel key construction — e.g. empty string produces ':ticket:'

**Suggestion:** Add guard: if not resource_type or not resource_id: await websocket.close(WS_1008_POLICY_VIOLATION, 'Invalid path'); return

File: tests/unit/test_websocket.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+from __future__ import annotations
+
+import os
+from unittest.mock import AsyncMock, MagicMock
+
+import pytest
+from starlette.websockets import WebSocketDisconnect
+
+_TEST_SECRET = "test-secret-32-chars-minimum-xx"  # noqa: S105
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:11]: **INFO** (suggestion)

**Issue:** Setting os.environ at module load time is a side-effect at import. Rule 131: async code must not hide blocking work; rule 16: comments explain 'why', not 'what'.

**Suggestion:** Move JWT_SECRET_KEY setup into a pytest hook or use pytest-env / monkeypatch in a session-scoped fixture instead of module-level mutation.

File: tests/unit/test_websocket.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+os.environ["JWT_SECRET_KEY"] = _TEST_SECRET
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py]: **WARNING** (suggestion)

**Issue:** Module-level os.environ mutation is invisible to any code that already cached JWT_SECRET_KEY before this module imported (e.g. conftest.py loading dotenv).

**Suggestion:** Use pytest.fixture with autouse=True and scope='session' that calls monkeypatch.setenv, or set JWT_SECRET_KEY in a pytest.ini [pytest:env] section via pytest-env.

File: tests/unit/test_websocket.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+# -------------------------------------------------------------------------- #
+
+
+class TestVerifyWsToken:
+    """Tests for verify_ws_token()."""
+
+    @pytest.fixture
+    def auth_service(self):
+        from src.services.auth_service import AuthService
+
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:34]: **WARNING** (suggestion)

**Issue:** Bare except: catches HTTPException and all subclasses — _decode_jwt raises HTTPException on expired/invalid tokens but verify_ws_token silently converts it to None, defeating the distinction between auth failure types.

**Suggestion:** Replace bare except Exception with except (jwt.ExpiredSignatureError, jwt.InvalidTokenError): so the two failure modes are preserved and distinguishable in tests.

File: tests/unit/test_websocket.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        return AuthService(MagicMock(), secret_key=_TEST_SECRET)
+
+    def test_valid_token_returns_context(self, auth_service):
+        """A valid JWT from query params should return an AuthContext with tenant_id."""
+        from src.api.routers.websocket import verify_ws_token
+
+        token = auth_service.generate_token(user_id=42, username="alice", role="admin", tenant_id=1)
+        ws = MagicMock()
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:42]: **WARNING** (suggestion)

**Issue:** Bare except: catches SystemExit, KeyboardInterrupt, and asyncio.CancelledError — the latter can propagate from cancelled tasks and should not be silently swallowed.

**Suggestion:** Add except (asyncio.CancelledError, KeyboardInterrupt, SystemExit): raise before the generic except Exception block.

File: tests/unit/test_websocket.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        """A malformed JWT should return None (not raise)."""
+        from src.api.routers.websocket import verify_ws_token
+
+        ws = MagicMock()
+        ws.query_params.get = MagicMock(return_value="not.valid.jwt")
+        ws.headers.get = MagicMock(return_value="")
+
+        ctx = verify_ws_token(ws)
+
+        assert ctx is None
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:75]: **INFO** (suggestion)

**Issue:** Test names are clear and describe intent per rule 67. No issues here.

**Suggestion:** No change needed.

File: tests/unit/test_websocket.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+    @pytest.fixture
+    def mock_manager(self):
+        """Return an AsyncMock that shadows _connection_manager for the current test."""
+        mock = MagicMock()
+        mock.join = AsyncMock()
+        mock.leave = AsyncMock()
+        return mock
+
+    async def test_valid_token_accepts_and_joins(self, valid_token, mock_manager):
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py]: **INFO** (suggestion)

**Issue:** Tests use try/finally to restore module-level state after patching, which is correct but manual and easy to forget in new tests.

**Suggestion:** Consider extracting a pytest fixture that uses request.addfinalizer or a context manager to automate unpatching, reducing the risk of state leakage between tests.

File: tests/unit/test_websocket.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+            from src.api.routers.websocket import ws_channel
+
+            ws = AsyncMock()
+            ws.query_params.get = MagicMock(return_value=valid_token)
+            ws.headers.get = MagicMock(return_value="")
+            ws.accept = AsyncMock()
+            ws.close = AsyncMock()
+            ws.receive_text = AsyncMock(side_effect=WebSocketDisconnect())
+
+            await ws_channel(ws, "ticket", "42")
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py]: **INFO** (suggestion)

**Issue:** test_valid_token_accepts_and_joins validates the channel key format correctly ('1:ticket:42'), assertion at line 134 checks tenant_id derived from the token, which correctly exercises the multi-tenancy path.

**Suggestion:** No change needed.

**Comment by yihua-zhuo** [comment]
## Claude Code Review (inline)

**Status:** `fail`
**Inline comments posted:** 8

**Summary:** Reviewed 5 file(s) between origin/master and origin/auto/issue-503.

### Issues without a usable file/line
- `src/api/routers/__init__.py`: Minimal init file correctly re-exports the websocket router with no issues.

### Issues that could not be posted inline
`src/internal/middleware/fastapi_auth.py` L20: 
**WARNING** (suggestion)

**Issue:** Rule 8 / Rule 129: `get_jwt_secret()` imports `main` inside the function body to access `app.state.jwt_secret`. If `main.py` imports anything from this module at startup (e.g., a router that re-exports via fastapi_auth), the import can fail with a circular dependency.

**Suggestion:** Avoid runtime import of `main`. Pass the secret in via a settings/data-class injected at app startup (e.g., a `AppState` dataclass set on `request.app.state` and accessed via `request.app.state.jwt_secret`), or rely solely on the environment variable path. — _HTTP 422: {"message":"Validation Failed","errors":[{"resource":"PullRequestReviewComment","code":"custom","field":"pull_request_review_thread.line","message":"could not be resolved"}],"documentation_url":"https://docs.github.com/rest/pulls/comments#create-a-review-comment-for-a-pull-request","status":"422"}_
`src/internal/middleware/fastapi_auth.py` L31: 
**INFO** (suggestion)

**Issue:** Rule 8 / Rule 28: `os.environ.get` is evaluated at module load time (lines 11 and 31). The production guard check runs once when the module is first imported, not lazily — this could pass in tests that set the env var after import but before production paths are hit.

**Suggestion:** Move the production guard into `get_jwt_secret()` or an explicit `verify_production_config()` call at startup rather than at module-load time. — _HTTP 422: {"message":"Validation Failed","errors":[{"resource":"PullRequestReviewComment","code":"custom","field":"pull_request_review_thread.line","message":"could not be resolved"}],"documentation_url":"https://docs.github.com/rest/pulls/comments#create-a-review-comment-for-a-pull-request","status":"422"}_

File: src/api/routers/websocket.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
+"""WebSocket router — JWT-authenticated channel subscription."""
+
+from fastapi import APIRouter, HTTPException, WebSocket, WebSocketDisconnect
```
Comment (yihua-zhuo) [inline src/api/routers/websocket.py]: **WARNING** (suggestion)

**Issue:** HTTPException is imported but never referenced — only the raised HTTPException from _decode_jwt is caught via the bare `except HTTPException:` which does not need the name

**Suggestion:** Remove `HTTPException` from the import line: `from fastapi import APIRouter, WebSocket, WebSocketDisconnect`

File: tests/unit/test_websocket.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        ws.receive_text = AsyncMock(side_effect=WebSocketDisconnect())
+
+        await ws_channel(ws, "ticket", "42")
+
+        ws.accept.assert_called_once()
+        patch_connection_manager.join.assert_called_once()
+        assert patch_connection_manager.join.call_args[0][0] == "1:ticket:42"
+        patch_connection_manager.leave.assert_called_once()
+
+    async def test_missing_token_closes_connection(self, patch_connection_manager):
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:131]: **INFO** (suggestion)

**Issue:** Cross-tenant negative cases are not tested for ws_channel: no test verifies that a token for tenant A cannot join a channel owned by tenant B.

**Suggestion:** Add a test that generates a token for tenant_id=2 and asserts it cannot join channel constructed from tenant_id=1 (e.g. by checking channel name format or ConnectionManager authorization logic).

File: .plans/issue-503.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+Refactor the JWT decode logic from `require_auth` into a standalone `_decode_jwt(token: str) -> AuthContext` function that raises `HTTPException(401)` on any decode error. Both `require_auth` and the upcoming WebSocket router will call this function.
+
+```python
+def _decode_jwt(token: str) -> AuthContext:
+    secret = get_jwt_secret()
+    try:
+        payload = jwt.decode(token, secret, algorithms=[JWT_ALGORITHM], options={"verify_aud": False})
+    except jwt.ExpiredSignatureError:
+        raise HTTPException(status_code=401, detail="Token 已过期")
```
Comment (yihua-zhuo) [inline .plans/issue-503.md:36]: **CRITICAL** (critical)

**Issue:** The `raise HTTPException` statement for `jwt.InvalidTokenError` is missing its closing quote and parenthesis — the string `'无效的 Token'` is unterminated, making this invalid Python.

**Suggestion:** Change `raise HTTPException(status_code=401, detail="无效的 Token` to `raise HTTPException(status_code=401, detail="无效的 Token")` — close the string and the constructor call.

File: tests/unit/test_websocket.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+# -------------------------------------------------------------------------- #
+# TestVerifyWsToken — test verify_ws_token() in isolation
+# -------------------------------------------------------------------------- #
+
+
+class TestVerifyWsToken:
+    """Tests for verify_ws_token()."""
+
+    @pytest.fixture
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:31]: **CRITICAL** (critical)

**Issue:** auth_service fixture passes secret_key=_TEST_SECRET but verify_ws_token -> _decode_jwt calls get_jwt_secret() which reads os.environ — if fastapi_auth.py imports before set_jwt_secret runs, secrets differ and tokens silently produce None from _decode_jwt, making the entire test suite a false pass.

**Suggestion:** Add a module-level autouse fixture (or add to conftest.py) that calls monkeypatch.setenv('JWT_SECRET_KEY', _TEST_SECRET, prepend=False) before any modules import fastapi_auth. Alternatively, patch get_jwt_secret directly in each affected test so _decode_jwt resolves the test secret.

File: src/api/routers/websocket.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        auth_header = websocket.headers.get("authorization", "")
+        if auth_header.startswith("Bearer "):
+            raw_token = auth_header[7:]
+
+    if not raw_token:
+        return None
+
+    try:
+        return _decode_jwt(raw_token)
+    except Exception:
```
Comment (yihua-zhuo) [inline src/api/routers/websocket.py:34]: **WARNING** (suggestion)

**Issue:** Rule 25: `except Exception` is a bare catch-all that also intercepts `KeyboardInterrupt` and `SystemExit`; the plan explicitly specifies catching `HTTPException` from `_decode_jwt`.

**Suggestion:** Replace `except Exception` with `except HTTPException` to match the plan's intent and avoid masking unexpected runtime errors.

File: src/api/routers/websocket.py [hunk 17]
```diff
@@ -0,0 +1,68 @@
+"""WebSocket router — JWT-authenticated channel subscription."""
+
+from fastapi import APIRouter, WebSocket, WebSocketDisconnect
+from starlette.status import WS_1008_POLICY_VIOLATION
+
+from internal.middleware.fastapi_auth import AuthContext, _decode_jwt
```
Comment (yihua-zhuo) [inline src/api/routers/websocket.py:6]: **INFO** (suggestion)

**Issue:** Rule 16: `_decode_jwt` is a private module function (not in `__all__`) — importing it here couples the router to an internal implementation detail of fastapi_auth.py.

**Suggestion:** Consider making `_decode_jwt` the public interface (e.g., export it from fastapi_auth.py) or wrapping it in a stable, versioned API. No change required if the contract is intentionally internal.

File: tests/unit/test_websocket.py [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+        ws.accept = AsyncMock()
+        ws.close = AsyncMock()
+
+        await ws_channel(ws, "ticket", "99")
+
+        ws.close.assert_called_once_with(code=1008, reason="Unauthorized")
+        ws.accept.assert_not_called()
+        patch_connection_manager.join.assert_not_called()
+
+    async def test_cross_tenant_cannot_join_other_tenant_channel(self, auth_service, patch_connection_manager):
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:147]: **WARNING** (suggestion)

**Issue:** test_cross_tenant_cannot_join_other_tenant_channel does not actually test cross-tenant blocking — it only verifies that a tenant-2 token produces a tenant-2-keyed channel name; it never verifies that tenant 1 data is inaccessible to tenant 2.

**Suggestion:** Add a test that generates a tenant-1 token (tenant_id=1), calls ws_channel, accepts it, then asserts that _connection_manager.join was called with channel '1:ticket:42' — the pairing of positive (tenant 2 joins 2:*) and negative (tenant 1 is isolated) assertions is required per rule 126.

File: tests/unit/test_websocket.py [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+        ctx = verify_ws_token(ws)
+
+        assert ctx is not None
+        assert ctx.user_id == 42
+        assert ctx.tenant_id == 1
+        assert "admin" in ctx.roles
+
+    def test_missing_token_returns_none(self):
+        """When neither query param nor header provides a token, return None."""
+        from src.api.routers.websocket import verify_ws_token
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:55]: **INFO** (suggestion)

**Issue:** resource_type and resource_id are path parameters typed as str with no explicit validation in ws_channel beyond the falsy/None guards (line 55); a tenant could technically broadcast to an empty-string channel that passes the guard but collides with or bypasses namespace expectations.

**Suggestion:** Add explicit assertion(s) in the test suite that ws_channel raises or closes on malformed resource_type/resource_id values such as whitespace-only strings or excessively long IDs.

File: tests/unit/test_websocket.py [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+    def auth_service(self):
+        from src.services.auth_service import AuthService
+
+        return AuthService(MagicMock(), secret_key=_TEST_SECRET)
+
+    @pytest.fixture
+    def valid_token(self, auth_service):
+        return auth_service.generate_token(user_id=1, username="alice", role="admin", tenant_id=1)
+
+    async def test_valid_token_accepts_and_joins(self, valid_token, patch_connection_manager):
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:113]: **INFO** (suggestion)

**Issue:** Tests assert ws.accept and manager.join/leave are called but do not verify the manager.leave call uses the same channel and websocket as join, leaving a potential bug where leave is called with wrong arguments undetected.

**Suggestion:** Add asserts on patch_connection_manager.leave.call_args to confirm channel value matches the join channel and the websocket argument is the same object that was passed to join.

File: tests/unit/test_websocket.py [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+from unittest.mock import AsyncMock, MagicMock
+
+import pytest
+from starlette.status import WS_1008_POLICY_VIOLATION
+from starlette.websockets import WebSocketDisconnect
+
+_TEST_SECRET = "test-secret-32-chars-minimum-xx"  # noqa: S105
+
+
+@pytest.fixture(scope="session", autouse=True)
```
Comment (yihua-zhuo) [inline tests/unit/test_websocket.py:14]: **INFO** (suggestion)

**Issue:** set_jwt_secret uses scope='session' autouse=True with os.environ mutate/deletes; if any test raises before yield, the cleanup del os.environ['JWT_SECRET_KEY'] never runs and subsequent tests in the same session inherit a stale or absent env var.

**Suggestion:** Use pytest's monkeypatch fixture with scope='session' instead of manual os.environ mutation; or wrap in try/finally to guarantee cleanup even on early-exit exceptions.

**Comment by yihua-zhuo** [comment]
**Feature scope appears incomplete vs #503** ⚠️

The PR's intended files and commits are not present in the local repository — no websocket.py, no test_websocket.py, and no fastapi-migration branch with the implementation exists locally.

**Missing:**
- src/api/routers/websocket.py does not exist
- tests/unit/test_websocket.py does not exist
- src/api/routers/__init__.py has no websocket router export
- src/internal/middleware/fastapi_auth.py is unchanged
- No branch containing the PR commits is accessible locally or via remote

_Posted by the PR-reviewer agent. Push more commits and the next tick will re-evaluate._
